danascheider/skyrim_inventory_management_frontend
# Test shopping list editing functionality

## Context

[**Test shopping list editing functionality**](https://trello.com/c/4gFa4SER/110-test-shopping-list-editing-functionality)
[**Add test cases for changing shopping list title**](https://trello.com/c/CnvFITTL/83-add-test-cases-for-changing-shopping-list-title)

After the refactor to add the Game model, we had to make some changes to the shopping list page. The shopping list edit functionality (changing the title of the shopping list) should still be working since the backend routes haven't changed, but there are no tests for this and the context where the API call is made could use a little refactoring.

## Changes

* Add test cases in Jest for editing a shopping list title
* Refactor shopping list page test suite into three different files
* Create `titlecase` util that uses the same logic the backend uses to correct casing in shopping list titles

## Considerations

The test cases for creating a shopping list were leaking into the test cases for editing a shopping list, which turned out to be caused by a [possible bug](https://github.com/testing-library/react-testing-library/issues/716) in the React testing library. Because the test file was getting a bit unwieldy anyway, I ended up splitting up the shopping list page tests into three different files - one for fetching shopping lists, one for creating one, and one for editing.

I also took the opportunity to refactor the `performShoppingListUpdate` function into something more similar to what we're using in the `GamesProvider` - it now takes a `callbacks` object with a separate callback for each possible API response (the `onInternalServerError` callback also runs if there is an unexpected error while processing the API response).

Although testing this proved impractical, the testing card above (the second one linked) indicates that editing functionality should be changed so that the front end also title-cases the title the user enters. This prevents the title from flashing what the user entered and then changing to title case when the API response comes back. For this purpose, I added the `titlecase` function, which uses the identical logic to the backend's `Titlecase` lib module, to make sure the new title is title-cased before it's updated in the view.

## Manual Test Cases

Testing the API's error responses will involve programming your development API to return the error codes in question. This PR doesn't actually change functionality but rather fixes it so it is as it was before. For that reason, you should only need to test what happens with each possible status code from the API. Note that a 405 error, which is returned if the requested list is an aggregate list, should never be returned since there is no edit icon or form for aggregate lists.

## Screenshots and GIFs

### 404 Error

#### 1201px

![issue-110_404-error_1201px](https://user-images.githubusercontent.com/5115928/126886099-0945cffb-db8e-414f-a3b6-e1eadab371eb.png)

#### 1025px

![issue-110_404-error_1025px](https://user-images.githubusercontent.com/5115928/126886101-daa6b63b-0322-4c1b-9264-aac3a47192c1.png)

#### 769px

![issue-110_404-error_769px](https://user-images.githubusercontent.com/5115928/126886106-b4d05490-e59d-4df5-8f1f-389d153af296.png)

#### 601px

![issue-110_404-error_601px](https://user-images.githubusercontent.com/5115928/126886108-cf35d522-f4c5-455e-96b2-b6bdf73d9afd.png)

#### 425px

![issue-110_404-error_425px](https://user-images.githubusercontent.com/5115928/126886110-576dc49f-00ae-498b-b25e-469f4554eacf.png)

### 422 Error

#### 1201px

![issue-110_422-error_1201px](https://user-images.githubusercontent.com/5115928/126886113-0bea41c0-8c88-4433-8c5e-b251956e7e82.png)

#### 1025px

![issue-110_422-error_1025px](https://user-images.githubusercontent.com/5115928/126886120-81f41d4b-381f-4a70-87d1-e3ff6e63e017.png)

#### 769px

![issue-110_422-error_769px](https://user-images.githubusercontent.com/5115928/126886123-655e44d3-70b3-465e-b787-dfe71dd90419.png)

#### 601px

![issue-110_422-error_601px](https://user-images.githubusercontent.com/5115928/126886126-acf4baa3-cb51-476d-90ff-2e22f55121dc.png)

#### 425px

![issue-110_422-error_425px](https://user-images.githubusercontent.com/5115928/126886131-6124c21a-b0af-4037-b8de-494ce329a6f1.png)

### 500 or Other Error

#### 1201px

![issue-110_500-error_1201px](https://user-images.githubusercontent.com/5115928/126886134-05fcc396-6be1-4524-b505-9bf5952742cf.png)

#### 1025px

![issue-110_500-error_1025px](https://user-images.githubusercontent.com/5115928/126886138-879a22b9-4be7-47ee-a624-98a87791afd4.png)

#### 769px

![issue-110_500-error_769px](https://user-images.githubusercontent.com/5115928/126886145-f1ad91d4-4fc5-4ae8-afb0-fa78e0ea6ae8.png)

#### 601px

![issue-110_500-error_601px](https://user-images.githubusercontent.com/5115928/126886146-71a5ae98-12f8-42e5-aa56-0e975f84246d.png)

#### 425px

![issue-110_500-error_425px](https://user-images.githubusercontent.com/5115928/126886149-6b3c9a31-d54d-4e9c-9350-2b2f3969f3fe.png)
